app/services/order_detail_batch_updater.rb

```diff
@@ -0,0 +1,144 @@
+  #
+  #                                              'unassign'
+  #                                              (to unassign current user)
+  #
+  #
+  #   :order_status_id        integer or string: id of an OrderStatus
+  #                                              they should be set to
+  #
+  #
+  # Optional Parameters:
+  #
+  # msg_type:         a plural string used in error/success messages to indicate
+  #                   type of records,
+  #                   (since this class method is also used to update
+  #                   order_details associated with reservations)
+  #                   defaults to 'orders'
+
+  def self.permitted_attributes
+    @permitted_attributes ||= %i(assigned_user_id order_status_id)
+  end
+
+  def initialize(order_detail_ids, facility, user, params, msg_type = "orders")
+    @order_detail_ids = order_detail_ids
+    @facility = facility
+    @user = user
+    @params = params
+    @msg_type = msg_type
+    @msg_hash = {}
+  end
+
+  def update!
+    unless order_detail_ids.present?
+      msg_hash[:error] = "No #{msg_type} selected"
+      return msg_hash
+    end
+
+    if order_details.any? { |od| !(od.state.include?("inprocess") || od.state.include?("new")) }
+      msg_hash[:error] = "There was an error updating the selected #{msg_type}"
+      return msg_hash
+    end
+
+    OrderDetail.transaction do
+      update_order_details_from_params
+
+      unless changes?
+        msg_hash[:notice] = "No changes were required"
+        return msg_hash
+      end
+
+      begin
+        order_details.all?(&:save!)
+        msg_hash[:notice] = "The #{msg_type} were successfully updated"
+      rescue
+        msg_hash[:error] = "There was an error updating the selected #{msg_type}"
+        raise ActiveRecord::Rollback
+      end
+    end
+
+    msg_hash
+  end
+
+  private
+
+  def changes?
+    @changes.present?
+  end
+
+  def flag_change
+    @changes = true
+  end
+
+  def order_details
+    @order_details ||= facility.order_details.where(id: order_detail_ids)
+  end
+
+  def order_status
+    @order_status ||= OrderStatus.find(params[:order_status_id])
+  end
+
+  def update_order_details_from_params
+    self.class.permitted_attributes.each do |attribute|
+      next if params[attribute].blank?
+
+      flag_change
+
+      if attribute == :order_status_id
+        update_all_order_status_ids
+      else
+        value = params[attribute] == "unassign" ? nil : params[attribute]
+        update_all(attribute, value)
+      end
+    end
+  end
+
+  def update_all(attribute, value)
+    order_details.each do |order_detail|
+      order_detail.public_send("#{attribute}=", value)
+    end
+  end
+
+  def update_all_order_status_ids
+    order_details.each do |order_detail|
+      # cancel reservation order details
+      if order_status.id == OrderStatus.canceled.first.id && order_detail.reservation
+        unless order_detail.cancel_reservation(user, order_status, true)
```

Comment:
I don't know if it's worth doing now, but `OrderDetail#update_order_status!` is essentially the same as `change_status!` but will handle the reservation cancelation.